src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,449 @@
+#include <cerrno>
+#include <cctype>
+#include <libgen.h>
+
+#include <mono/metadata/assembly.h>
+
+#include "embedded-assemblies.hh"
+#include "cpp-util.hh"
+#include "globals.hh"
+
+using namespace xamarin::android::internal;
+
+void
+EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_should_register should_register)
+{
+	uint32_t cd_offset;
+	uint32_t cd_size;
+	uint16_t cd_entries;
+
+	if (!zip_read_cd_info (fd, cd_offset, cd_size, cd_entries)) {
+		log_fatal (LOG_DEFAULT,  "Failed to read the EOCD record from APK file %s", apk_name);
+		exit (FATAL_EXIT_NO_ASSEMBLIES);
+	}
+#ifdef DEBUG
+	log_info (LOG_DEFAULT, "Central directory offset: %u", cd_offset);
+	log_info (LOG_DEFAULT, "Central directory size: %u", cd_size);
+	log_info (LOG_DEFAULT, "Central directory entries: %u", cd_entries);
+#endif
+	off_t result = ::lseek (fd, static_cast<off_t>(cd_offset), SEEK_SET);
+	if (result < 0) {
+		log_fatal (LOG_DEFAULT, "Failed to seek to central directory position in the APK file %s. %s (result: %d; errno: %d)", apk_name, std::strerror (errno), result, errno);
+		exit (FATAL_EXIT_NO_ASSEMBLIES);
+	}
+
+	// C++17 allows template parameter type inference, but alas, Apple's antiquated compiler does
+	// not support this particular part of the spec...
+	simple_pointer_guard<uint8_t[]>  buf (new uint8_t[cd_size]);
+	const char           *prefix     = get_assemblies_prefix ();
+	size_t                prefix_len = strlen (prefix);
+	size_t                buf_offset = 0;
+	uint16_t              compression_method;
+	uint32_t              local_header_offset;
+	uint32_t              data_offset;
+	uint32_t              file_size;
+	char                 *entry_name;
+	char                 *file_name;
+
+	ssize_t nread = read (fd, buf.get (), cd_size);
+	if (static_cast<size_t>(nread) != cd_size) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory from the APK archive %s. %s (nread: %d; errno: %d)", apk_name, std::strerror (errno), nread, errno);
```

Comment:
Given that all of these are exiting with `FATAL_EXIT_NO_ASSEMBLIES`, would it make more sense for all of the `log_*()` invocations to use `LOG_ASSEMBLY`?